api/api/views/media_views.py

```diff
@@ -42,7 +43,20 @@ def __init__(self, *args, **kwargs):
             raise ValueError(msg)
 
     def get_queryset(self):
-        return self.model_class.objects.all()
+        # The alternative to a sub-query would be using `extra` to do a join
+        # to the content provider table and filtering `filter_content`. However,
+        # that assumes that a content provider entry exists, which is not necessarily
+        # the case. We often don't add a content provider until after works from
+        # new providers are available in the API, and sometimes not even then.
+        # Search returns results with providers that do not have a ContentProvider
+        # table entry. Therefore, to maintain that assumption, a subquery is the only
+        # workable approach, as Django's `extra` does not provide any facility for
+        # handling null relations on the join.
+        return self.model_class.objects.exclude(
+            provider__in=ContentProvider.objects.filter(
+                filter_content=True
+            ).values_list("provider_identifier")
```

Comment:
I am not 100% sure of the performance implications but would it be faster if we get this list outside and use `.all` instead of `.exclude` if the list is empty?

Reply:
Good question! For it to matter the list would need to be empty. In production it currently is not and I don't necessarily foresee it being empty unless we do some significant work to remove certain providers.

In the interest of checking for it anyway, though, I ran explain locally on both scenarios.

- Here's the explain for when the subquery is empty: https://explain.depesz.com/s/WD8Y
- Here's the explain for when the subquery is not empty: https://explain.depesz.com/s/5Q4O
- Here's the explain with no subquery at all: https://explain.depesz.com/s/FPWt

I ran these locally and I currently have no providers with `content_filter=true`, which is why the first query would have an empty subquery (you can see the filter on the subquery removed 4 rows, i.e., all four content providers I have locally).

Comparing both, you can see that the empty subquery results in a faster query, which makes sense, Postgres has nothing to compare the rows to. However, it is indeed slower than no subquery at all, which also makes sense as it isn't obvious to me that Postgres is _skipping_ the "in" check (though it sure seems like it should be smart enough to do so :shrug:).

Additionally, it is also simpler and faster (it seems) to use a literal list for the right side of "in" vs a subquery, though I'm not sure why. At this point we'd also have to factor in Django ORM overhead. Does the overhead of executing a separate query (meaning the new request to the database) then serializing it into Python objects, then passing those to Django ORM for it to then serialize back into the string format for Postgres (which includes escaping the input) compensate for the better query time performance in Postgres? I'd certainly be curious to know, but to get meaningful data I think we would need to run both in production and compare how they go. It's hard to tell locally what the holistic performance difference would be, especially when we know that in production the case where we can remove the provider check simply doesn't exist because we do have filtered providers in production.